sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -301,12 +306,139 @@ case class GpuCaseWhen(
+      branches.foldRight[Any](elseRet) {
+        case ((predicateExpr, trueExpr), falseRet) =>
+          computeIfElse(batch, predicateExpr, trueExpr, falseRet)
+        }
+      }
+    }
+
+  /**
+   * Perform lazy evaluation of each branch sa that we only evaluate the THEN expressions
+   * against rows where the WHEN expression is true.
+   */
+  private def columnarEvalWithSideEffects(batch: ColumnarBatch): Any = {
+    val colTypes = GpuColumnVector.extractTypes(batch)
+
+    // track cumulative state of predicate evaluation per row so that we never evaluate expressions
+    // for a row if an earlier expression has already been evaluated to true for that row
+    var cumulativePred: Option[GpuColumnVector] = None
+
+    // this variable contains the currently evaluated value for each row and gets updated
+    // as each branch is evaluated
+    var currentValue: Option[GpuColumnVector] = None
+
+    try {
+      withResource(GpuColumnVector.from(batch)) { tbl =>
+
+        // iterate over the WHEN THEN branches first
+        branches.foreach {
+          case (whenExpr, thenExpr) =>
+            // evaluate the WHEN predicate
+            withResource(GpuExpressionsUtils.columnarEvalToColumn(whenExpr, batch)) { whenBool =>
+              // we only want to evaluate where this WHEN is true and no previous WHEN has been true
+              val whenBoolAdjusted: GpuColumnVector = cumulativePred match {
+                case Some(prev) =>
+                  withResource(prev.getBase.not()) { notPreviouslyTrue =>
+                    GpuColumnVector.from(
+                      whenBool.getBase.and(notPreviouslyTrue), DataTypes.BooleanType)
+                  }
+                case None =>
+                  whenBool.incRefCount()
+              }
+
+              withResource(whenBoolAdjusted) { _ =>
+                if (isAllTrue(whenBoolAdjusted)) {
+                  // if this WHEN predicate is true for all rows and no previous predicate has
+                  // been true then we can return immediately
+                  return GpuExpressionsUtils.columnarEvalToColumn(thenExpr, batch)
+                }
+                val thenValues = withResource(filterBatch(tbl,
+                    whenBoolAdjusted.getBase, colTypes)) {
+                  trueBatch => GpuExpressionsUtils.columnarEvalToColumn(thenExpr, trueBatch)
+                }
+                withResource(thenValues) { _ =>
+                  withResource(gather(whenBoolAdjusted.getBase, thenValues)) { gather =>
+                    currentValue = Some(currentValue match {
+                      case Some(v) =>
+                        withResource(v) { _ =>
+                          GpuColumnVector.from(whenBoolAdjusted.getBase.ifElse(gather, v.getBase),
+                            dataType)
+                        }
+                      case _ =>
+                        GpuColumnVector.from(gather.incRefCount(), dataType)
+                    })
+                  }
```

Comment:
Nit: I'm not a fan of putting a `match` directly in the parameter list of something -- it adds quite a bit of complexity to reading the line.  I think this could be a good candidate for refactoring, as there's a core piece here that is useful and easy to understand in isolation, something like the following (my names probably could be better)
```scala
private def updateCumulativeResult(
    currentValue: Option[GpuColumnVector],
    thenValues: GpuColumnVector,
    gatherMask: ColumnVector): GpuColumnVector = {
  withResource(gather(gatherMask, thenValues)) { gather =>
    currentValue = Some(currentValue match {
      case Some(v) =>
        withResource(v) { _ =>
          GpuColumnVector.from(gatherMask.ifElse(gather, v.getBase), dataType)
        }
      case _ =>
        GpuColumnVector.from(gather.incRefCount(), dataType)
    }
  }
}

[...]

currentValue = Some(updateCumulativeResult(currentValue, whenBoolAdjusted.getBase, thenValues))
```

Based on the `thenValues` lifetime comment, you may want this method to be responsible for closing the `thenValue` if refactoring in this way.